mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java

```diff
@@ -1,178 +1,190 @@
+
+    @Inject
+    public SimpleMessageSearchIndex(MessageMapperFactory messageMapperFactory, MailboxMapperFactory mailboxMapperFactory, TextExtractor textExtractor, AttachmentContentLoader attachmentContentLoader) {
+        this.messageMapperFactory = messageMapperFactory;
+        this.mailboxMapperFactory = mailboxMapperFactory;
+        this.textExtractor = textExtractor;
+        this.attachmentContentLoader = attachmentContentLoader;
+    }
+    
+    @Override
+    public EnumSet<SearchCapabilities> getSupportedCapabilities(EnumSet<MessageCapabilities> messageCapabilities) {
+        return EnumSet.of(SearchCapabilities.MultimailboxSearch,
+            SearchCapabilities.Text,
+            SearchCapabilities.Attachment,
+            SearchCapabilities.PartialEmailMatch,
+            SearchCapabilities.AttachmentFileName);
+    }
+    
+    /**
+     * Walks down the query tree's conjunctions to find a UidCriterion
+     * @param crits - list of Criterion to search from
+     * @return
+     *      first UidCriterion found
+     *      null - if not found
+     */
+    private static UidCriterion findConjugatedUidCriterion(List<Criterion> crits) {
+        for (Criterion crit : crits) {
+            if (crit instanceof UidCriterion) {
+                return (UidCriterion) crit;
+            } else if (crit instanceof ConjunctionCriterion) {
+                return findConjugatedUidCriterion(((ConjunctionCriterion) crit)
+                        .getCriteria());
+            }
+        }
+        return null;
+    }
+    
+    @Override
+    public Stream<MessageUid> search(MailboxSession session, final Mailbox mailbox, SearchQuery query) throws MailboxException {
+        Preconditions.checkArgument(session != null, "'session' is mandatory");
+        return searchResults(session, Flux.just(mailbox), query)
+            .filter(searchResult -> searchResult.getMailboxId().equals(mailbox.getMailboxId()))
+            .map(SearchResult::getMessageUid)
+            .toStream();
+    }
+
+    private List<SearchResult> searchResults(MailboxSession session, Mailbox mailbox, SearchQuery query) throws MailboxException {
+        MessageMapper mapper = messageMapperFactory.getMessageMapper(session);
+
+        final SortedSet<MailboxMessage> hitSet = new TreeSet<>();
+
+        UidCriterion uidCrit = findConjugatedUidCriterion(query.getCriterias());
+        if (uidCrit != null) {
+            // if there is a conjugated uid range criterion in the query tree we can optimize by
+            // only fetching this uid range
+            UidRange[] ranges = uidCrit.getOperator().getRange();
+            for (UidRange r : ranges) {
+                Iterator<MailboxMessage> it = mapper.findInMailbox(mailbox, MessageRange.range(r.getLowValue(), r.getHighValue()), FetchType.Metadata, UNLIMITED);
+                while (it.hasNext()) {
+                    hitSet.add(it.next());
+                }
+            }
```

Comment:
use streams

Reply:
Legacy code, nothing to do with my changes